README.md
# [nosh](https://www.merriam-webster.com/dictionary/nosh)

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/nosh)](https://CRAN.R-project.org/package=nosh)
[![R-CMD-check](https://github.com/seankross/nosh/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/seankross/nosh/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

Opinionated, lightweight, read-only Notion API access.

## Installation

You can install the latest development version:

``` r
remotes::install_github("seankross/nosh@main")
```

## Getting Started

To use Nosh you need to get your own Internal Integration Secret (IIS) from Notion
which is specific to your integration. To do this go to
[Notion Integrations](https://www.notion.so/my-integrations) and create a new
integration. Go to the **Secrets** page of your new integration and you should
see the IIS there. To make R aware of your IIS you need to set the environmental
variable `NOTION_IIS`. You can do this in your `.Renviron` or with this line of
code:

```r
Sys.setenv("NOTION_IIS" = "secret_5nLEIYqDd6o58gd1ZAPP4OMH9OnVozyhN9XQysN1nFE")
```

Your Notion integration is only aware of the pages and databases that you
explicitly give it access to. 

- To grant your integration access to **User Capabilities**, navigate to the **Capabilities** page (located below **Secrets**) and select "Read user information including email addresses" within the "User capabilities" subsection.
- To give your integration access to a page or
database, go to that page or database, select the three dot icon in the 
upper-right, click **Add connections**, then find the name of your Notion
integration and click on it. Your integration (and R) is now able to access
that page or database.


Nosh has three main functions:

``` r
library(nosh)

# List all users
nosh::users()
# If you get HTTP 403 Forbidden Error,
# grant your integration access to user capabilities

# List all pages and databases
nosh::search_all()

# List only pages and databases that match a query
nosh::search_all("Customer DB")

# Get a database as you would see it in Notion as a data frame. First you need
# to get the specific database ID.
db_id <- nosh::search_all("Customer DB")$id

# Now get the database
nosh::databases(db_id)
```

## Strong opinions about this package:

1. This package will always only have read-only functionality.
2. Functions in this package will always only return data frames.
3. The number of functions and arguments in this package will always be small.

## Related work:

- https://github.com/Eflores89/notionR


NEWS.md
# nosh 0.1.0

* Initial release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(nosh)
library(httr2)
library(jsonlite)

test_check("nosh")


tests/testthat/test-users.R
test_that("A response from the Notion Users API returns a data frame.", {
  test_response <- response(
    url = "https://api.notion.com/v1/users?page_size=100",
    headers = "Content-Type: application/json",
    body = list(has_more = FALSE,
                results = list(list(
                  name = "Sean",
                  id = "ABC123",
                  type = "person"))) |>
      toJSON(auto_unbox = TRUE) |>
      as.character() |>
      charToRaw()
  )

  result <- httr2::with_mock(\(x) test_response, users())
  expect_true(is.data.frame(result))
  expect_true(nrow(result) > 0)
})


R/users.R
#' List all users
#'
#' Get a data frame with the names, ids, and types of all users associated with
#' your account.
#'
#' @importFrom httr2 request req_user_agent req_headers req_url_path_append req_perform resp_body_json req_url_query
#' @importFrom dplyr tibble
#' @importFrom purrr map_dfr
#' @export
#' @examples
#' \dontrun{
#' users()
#' }
users <- function() {
  rbj <- list()
  rbj_index <- 1

  req_temp <- request("https://api.notion.com") |>
    req_user_agent("Nosh (https://github.com/seankross/nosh)") |>
    req_headers("Content-Type" = "application/json") |>
    req_headers("Authorization" = Sys.getenv("NOTION_IIS")) |>
    req_headers("Notion-Version" = "2022-06-28") |>
    req_url_path_append("v1") |>
    req_url_path_append("users") |>
    req_url_query(page_size = 100)

  rbj[[rbj_index]] <- req_temp |>
    req_perform() |>
    resp_body_json()

  while (rbj[[rbj_index]]$has_more) {
    rbj_index <- rbj_index + 1
    rbj[[rbj_index]] <- req_temp |>
      req_url_query(start_cursor = rbj[[rbj_index - 1]]$next_cursor) |>
      req_perform() |>
      resp_body_json()
  }

  if(length(rbj) > 1) {
    rbj_results <- reduce(rbj, function(x, y) c(x$results, y$results))
  } else {
    rbj_results <- rbj[[1]]$results
  }

  rbj_results |>
    map_dfr(\(x) tibble(name = x$name,
                        id = x$id,
                        type = x$type))
}


R/search_all.R
#' Search for pages and databases
#'
#' Specify a query to search all of the Notion pages and databases that you have
#' access to.
#'
#' @param query A string to search pages and databases. An empty string returns
#' all pages and databases you have access to (this is the default).
#' @importFrom httr2 request req_headers req_url_path_append req_body_json req_perform resp_body_json
#' @importFrom purrr reduce map_dfr
#' @importFrom dplyr tibble
#' @export
#' @examples
#' \dontrun{
#' # Returns all pages and databases
#' search_all()
#'
#' # Returns only pages and databases matching the query
#' search_all("Customer DB")
#' }
search_all <- function(query = "") {
  rbj <- list()
  rbj_index <- 1

  req_temp <- request("https://api.notion.com") |>
    req_headers("Content-Type" = "application/json") |>
    req_headers("Authorization" = Sys.getenv("NOTION_IIS")) |>
    req_headers("Notion-Version" = "2022-06-28") |>
    req_url_path_append("v1") |>
    req_url_path_append("search") |>
    req_body_json(list(
      query = query,
      page_size = 100))

  rbj[[rbj_index]] <- req_temp |>
    req_perform() |>
    resp_body_json()

  while (rbj[[rbj_index]]$has_more) {
    rbj_index <- rbj_index + 1
    rbj[[rbj_index]] <- req_temp |>
      req_body_json(list(
        query = query,
        page_size = 100,
        start_cursor = rbj[[rbj_index - 1]]$next_cursor)) |>
      req_perform() |>
      resp_body_json()
  }

  if(length(rbj) > 1) {
    rbj_results <- reduce(rbj, function(x, y) c(x$results, y$results))
  } else {
    rbj_results <- rbj[[1]]$results
  }

  rbj_results |> map_dfr(\(x) tibble(object = x$object,
                                     id = x$id,
                                     property_names = list(names(x$properties)),
                                     properties = list(x$properties),
                                     url = x$url,
                                     created_time = x$created_time,
                                     last_edited_time = x$last_edited_time,
                                     created_by_id = x$created_by$id,
                                     last_edited_by_object = x$last_edited_by$object,
                                     last_edited_by_id = x$last_edited_by$id,
                                     cover = list(x$cover),
                                     icon = list(x$icon),
                                     parent_type = x$type,
                                     parent_database_id = x$database_id,
                                     archived = x$archived
  ))

}


R/handle_property.R
# Given an entire property object, return a string
#' @importFrom purrr map_chr
handle_property <- function(prop) {
  result <- NULL

  if(prop$type == "checkbox") {
    result <- prop$checkbox
  } else if(prop$type == "created_by") {
    result <- prop$id
  } else if(prop$type == "created_time") {
    result <- prop$created_time
  } else if(prop$type == "date") {
    result <- prop$date$start
  } else if(prop$type == "email") {
    result <- prop$email
  } else if(prop$type == "files") {
    result <- length(prop$files)
  } else if(prop$type == "formula") {
    result <- prop[[prop$type]]
  } else if(prop$type == "last_edited_by") {
    result <- prop$last_edited_by$id
  } else if(prop$type == "last_edited_time") {
    result <- prop$last_edited_time
  } else if(prop$type == "multi_select") {
    result <- prop$multi_select |> map_chr(\(x) x$name) |> sort() |> paste0(collapse = "-")
  } else if(prop$type == "number") {
    result <- prop$number
  } else if(prop$type == "people") {
    result <- prop$people |> map_chr(\(x) x$name) |> sort() |> paste0(collapse = "-")
  } else if(prop$type == "phone_number") {
    result <- prop$phone_number
  } else if(prop$type == "relation") {
    result <- prop$relation |> map_chr(\(x) x$id) |> sort() |> paste0(collapse = "_")
  } else if(prop$type == "rollup") {
    result <- prop[[prop$type]]
  } else if(prop$type == "rich_text") {
    result <- prop$rich_text |> map_chr(\(x) x$plain_text) |> paste0(collapse = "")
  } else if(prop$type == "select") {
    result <- prop$select$name
  } else if(prop$type == "status") {
    result <- prop$status$name
  } else if(prop$type == "title") {
    result <- prop$title |> map_chr(\(x) x$plain_text) |> paste0(collapse = "")
  } else if(prop$type == "url") {
    result <- prop$url
  }

  ifelse(is.null(result), NA, result)
}


R/databases.R
#' Get a database as a data frame
#'
#' Returns all of the properties in a database as a data frame.
#'
#' @param id The ID of an individual database
#' @importFrom httr2 request req_headers req_url_path_append req_body_json req_perform resp_body_json
#' @importFrom purrr reduce map_dfr map_chr compose
#' @export
#' @examples
#' \dontrun{
#' # Find the ID of a specific database
#' db_id <- search_all("Customer DB")$id
#'
#' # Get that database
#' databases(db_id)
#' }
databases <- function(id) {
  rbj <- list()
  rbj_index <- 1

  req_temp <- request("https://api.notion.com") |>
    req_headers("Content-Type" = "application/json") |>
    req_headers("Authorization" = Sys.getenv("NOTION_IIS")) |>
    req_headers("Notion-Version" = "2022-06-28") |>
    req_url_path_append("v1") |>
    req_url_path_append("databases") |>
    req_url_path_append(id) |>
    req_url_path_append("query") |>
    req_body_json(list(
      page_size = 100
    ))

  rbj[[rbj_index]] <- req_temp |>
    req_perform() |>
    resp_body_json()

  while (rbj[[rbj_index]]$has_more) {
    rbj_index <- rbj_index + 1
    rbj[[rbj_index]] <- req_temp |>
      req_body_json(list(
        page_size = 100,
        start_cursor = rbj[[rbj_index - 1]]$next_cursor)) |>
      req_perform() |>
      resp_body_json()
  }

  if(length(rbj) > 1) {
    rbj_results <- reduce(rbj, function(x, y) c(x$results, y$results))
  } else {
    rbj_results <- rbj[[1]]$results
  }

  rbj_results |> map_dfr(function(x){
    map_chr(x$properties, compose(as.character, handle_property)) |> rev()
  })
}
